HTTP headers: deprecating "(Want-)?Digest", adding "(Want-)?(Content-|Repr-)Digest"

jfrech opened the pull request:
Legacy description based off of [draft 7](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-07).

jfrech committed:
HTTP header "Digest": set status to "deprecated", fixed legacy description

Legacy description based off of [draft 7](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-07).

hamishwillee commented on files/en-us/web/http/headers/digest/index.md:
```diff
@@ -2,11 +2,15 @@
 browser-compat: http.headers.Digest
 ---
 
 {{HTTPSidebar}}
```
This is the way deprecation is normally added. We'll keep your note because it is more informative.
```suggestion
{{HTTPSidebar}}{{Deprecated_Header}}
```

hamishwillee commented:
Thanks very much. This is great but we'll want to do more. From the [current spec](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers#name-since-draft-ietf-httpbis-digest):

> Introduced `Repr-Digest` and `Want-Repr-Digest`, and deprecated `Digest` and `Want-Digest`. Use of Structured Fields. 1993, 1919

So we want to deprecate `Want-Digest` and add `Repr-Digest` and `Want-Repr-Digest`. Do you want to take a shot at that?

Note that there are a couple of macros we use to indicate deprecation - I added one above for the header, but in lists of headers and whatever you can use `{{Deprecated_Inline}}`.

If `Repr-Digest is a representation header we'll want to add it here too: https://developer.mozilla.org/en-US/docs/Glossary/Representation_header

@teoli2003 The deprecated macro should really be able to include the information in @jfrech warning block. Can we modify the macro to allow specification of what the deprecation cause/replacement is?

jfrech commented:
@hamishwillee In e. g. [`Sec-CH-UA-Full-Version`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-UA-Full-Version), deprecation replacement is indicated by a **Note**. I only used a **Warning** since e. g. [`Accept-CH-Lifetime`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-CH-Lifetime) notes a draft removal as a **Warning**.  
So a more specific way to note prior feature existence and/or recommended newer alternatives may indeed be nice to have.

jfrech committed:
deprecated "Want-Digest"

jfrech committed:
stubbed "Content-Digest", "Repr-Digest", "Want-Content-Digest", "Want-Repr-Digest"

jfrech committed:
fix: misunderstanding of `Content-`

Cf. https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-12#section-1.4
Cf. https://www.rfc-editor.org/rfc/rfc9110#section-8.3

jfrech commented:
@hamishwillee May I double-check my interpretation of the spec with you?  

I read it to claim the `Repr-` to be the *Selected Representation*, fully dependent on Content Negotiation which takes `Content-Type` into account. Does this mean `Content-Digest` is of no use whatsoever if the "actual message content" (cf. https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-12#section-2) is not selectable as a representation (e.&hairsp;g. MDN's source files only being selectable as HTML)?

jfrech committed:
fix: both "Digest" and "Want-Digest" were both request and response headers

jfrech committed:
fix: "(Want-)?(Content-|Repr-)Digest" are all both request and response headers

jfrech committed:
"(Content-|Repr-)Digest": intralinked

jfrech committed:
"Repr-Digest": added example

jfrech committed:
glossary "representation header": weaved in digest headers, revised

jfrech commented:
@hamishwillee Should we remove "Content-Location" from the [glossary](https://developer.mozilla.org/en-US/docs/Glossary/Representation_header)? After all, it does not describe the representation sent by the associated HTTP message.

jfrech committed:
glossary: `{{Deprecated_Inline}}`'d "(Want-)?Digest"

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,25 @@ slug: Glossary/Representation_header
 ---
 
-A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes the particular _representation_ of the resource sent in an HTTP message body.
+A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes the particular _representation_ of the target resource sent in an HTTP message body.
```
Why "target"? 
My thinking here is that a representation header may be present in both request and response messages. So you can have a representation header without any resource in the message body. 

```suggestion
A **representation header** is an {{glossary("HTTP_header", "HTTP header")}} that describes the particular _representation_ of a resource.
```

jfrech replied:
I also changed "the particular _representation_" to "one particular _representation_".

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,25 @@ slug: Glossary/Representation_header
+For example, the same data might be formatted as a particular media type such as XML or JSON, localized to a particular written language or geographical region, and/or compressed or otherwise encoded for transmission. The underlying resource is the same in each case, but its representation is different.  
 
-Clients specify the formats that they prefer to be sent during [content negotiation](/en-US/docs/Web/HTTP/Content_negotiation) (using `Accept-*` headers), and the representation headers tell the client the format of the _selected representation_ they actually received.
+Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms: E.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}} does *not* affect the representation, wheres e.&hairsp;g. {{HTTPHeader("Content-Range")}} does.  
```
```suggestion
Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms.
For example, HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}} does *not* affect the representation, whereas e.g. {{HTTPHeader("Content-Range")}} does.  
```

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,25 @@ slug: Glossary/Representation_header
 
 Representation headers may be present in both HTTP request and response messages.
-If sent as a response to a `HEAD` request, they describe the body content that _would_ be selected if the resource was actually requested.
+If sent as a response to a `HEAD` request, they describe the body content representation that _would_ be selected if the resource was requested with a `GET` request.
```
Given it says "representation headers may be present in both HTTP request and response messages, I wonder if we need  an example in a request?

jfrech replied:
Both a positive and negative example may make sense: something akin to:
### positive example
```http
PUT /api/transact HTTP/1.1
Want-Content-Digest: sha-256
Content-Type: text/json
...
```
```http
HTTP/1.1 201 Created
Content-Digest: sha-256=:2IBI7hQn83oTCgB3Z/6apOl91WGoctRfRj/F9gkvVo8=:
Content-Encoding: br
Repr-Digest: sha-256=:W8oN3H3CmE/CBpV6ZPNozV2AIDzzQpWL7CCOXyDyDzI=:
...
```
### negative example
```http
GET /api/last-transaction HTTP/1.1
Accept: text/json, text/xml
Content-Digest: sha-256=:125+ms4/wPK0vMKM3oy9j4zu8HmPTmr7YVla229fhFc=:
...
```
```http
HTTP/1.1 406 Not Acceptable
Content-Digest: sha-256=:2IBI7hQn83oTCgB3Z/6apOl91WGoctRfRj/F9gkvVo8=:
...
```

jfrech replied:
But I still wonder when it makes sense for the server to respond with `Want-Content-Digest`.

hamishwillee replied:
Yes, these kinds of examples make sense to me. And no, I don't understand when a server could ever response with Want-Content-Digest. I doubt its like those client hints where a server can specify a preference for future headers from the client.

If we can't prove it isn't a valid use case, I'd still just not cover this.

jfrech replied:
Added in 72ee7b6a3b59f2f2ae41b848046c092390a48fd3.

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,25 @@ slug: Glossary/Representation_header
+If sent as a response to a `HEAD` request, they describe the body content representation that _would_ be selected if the resource was requested with a `GET` request.
 
-Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, and {{HTTPHeader("Content-Location")}}.
+Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Location")}}, {{HTTPHeader("Content-Range")}}, {{Deprecated_Inline}}{{HTTPHeader("Digest")}} and {{HTTPHeader("Repr-Digest")}}.
```
```suggestion
Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Location")}}, {{HTTPHeader("Content-Range")}}, {{HTTPHeader("Repr-Digest")}}, and {{HTTPHeader("Digest")}} {{Deprecated_Inline}}.
```

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,25 @@ slug: Glossary/Representation_header
 - {{Glossary("Payload header")}}
 - {{glossary("Entity header")}}
-- {{HTTPHeader("Digest")}}/ {{HTTPHeader("Want-Digest")}}
+- {{Deprecated_Inline}}{{HTTPHeader("Digest")}}, {{Deprecated_Inline}}{{HTTPHeader("Want-Digest")}}
+- {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Repr-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}, {{HTTPHeader("Want-Repr-Digest")}}
```
Deprecated things are "less important" and should IMO go last.

```suggestion
- {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Repr-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}, {{HTTPHeader("Want-Repr-Digest")}}
- {{HTTPHeader("Digest")}} {{Deprecated_Inline}}, {{HTTPHeader("Want-Digest")}} {{Deprecated_Inline}}
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee committed:
Update index.md - remove empty spaces for link checker

hamishwillee commented:
> g. [`Sec-CH-UA-Full-Version`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-UA-Full-Version), deprecation replacement is indicated by a **Note**. I only used a **Warning** since e. g. [`Accept-CH-Lifetime`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-CH-Lifetime) notes a draft removal as a **Warning**.

Yes, it would be great to be able to include deprecation replacement info in the macro itself, allowing just one note to be used to indicate deprecation. Let's see if we can get traction here: https://github.com/mdn/yari/issues/8969.

I'm going to change to a note. This is purely my taste - there is no real rule as far as I know.

hamishwillee commented on files/en-us/web/http/headers/digest/index.md:
```diff
@@ -2,12 +2,16 @@
+{{HTTPSidebar}}{{Deprecated_Header}}
 
-The **`Digest`** response HTTP header provides a {{Glossary("digest")}} of the _selected representation_ of the requested resource.
+> **Warning:** This header was removed from the specification in [draft 8](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-08). Use {{HTTPHeader("Repr-Digest")}} instead. For `id-*` digest algorithms, use {{HTTPHeader("Content-Digest")}}.
```
```suggestion
> **Note:** This header was removed from the specification in [draft 8](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-08).
> Use {{HTTPHeader("Repr-Digest")}} instead.
> For `id-*` digest algorithms, use {{HTTPHeader("Content-Digest")}}.
```

hamishwillee committed:
Update files/en-us/web/http/headers/digest/index.md

hamishwillee commented on files/en-us/web/http/headers/want-digest/index.md:
```diff
@@ -2,12 +2,16 @@
+{{HTTPSidebar}}{{Deprecated_Header}}
 
-The **`Want-Digest`** HTTP header is primarily used in a request, to ask the server to provide a {{Glossary("digest")}} of the requested resource using the {{HTTPHeader("Digest")}} response header.
+> **Warning:** This header was removed from the specification in [draft 8](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-08). Use {{HTTPHeader("Want-Repr-Digest")}} instead. For `id-*` digest algorithms, use {{HTTPHeader("Want-Content-Digest")}}.
```
```suggestion
> **Note:** This header was removed from the specification in [draft 8](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-08). Use {{HTTPHeader("Want-Repr-Digest")}} instead. For `id-*` digest algorithms, use {{HTTPHeader("Want-Content-Digest")}}.
```

hamishwillee committed:
Update files/en-us/web/http/headers/want-digest/index.md

hamishwillee commented on files/en-us/web/http/headers/digest/index.md:
```diff
@@ -2,12 +2,18 @@
+> Use {{HTTPHeader("Repr-Digest")}} instead.
+> For `id-*` digest algorithms, use {{HTTPHeader("Content-Digest")}}.
+
+The **`Digest`** response or request HTTP header provides the other side with a {{Glossary("digest")}} of the _selected representation_ or _content_. It can be requested by using the {{HTTPHeader("Want-Digest")}} header.
```
The "or content" seems redundant. The content is a selected representation. Would this be better?

```suggestion
The **`Digest`** response or request HTTP header provides a {{Glossary("digest")}} of the _selected representation_ of the requested/provided resource.
It can be requested by using the {{HTTPHeader("Want-Digest")}} header.
```

jfrech replied:
I did have a think; see further down.

hamishwillee commented on files/en-us/web/http/headers/digest/index.md:
```diff
@@ -40,18 +46,20 @@ Digest: <digest-algorithm>=<digest-value>,<digest-algorithm>=<digest-value>
 - `<digest-value>`
-  - : The result of applying the digest algorithm to the resource representation and encoding the result.
-    The choice of digest algorithm also determines the encoding to use: for example `SHA-256` uses base64 encoding.
+  - : The result of applying the digest algorithm to the resource representation and encoding (for non-`id-*` digest algorithm values) the result.
```
```suggestion
  - : The result of applying the digest algorithm to the resource representation and encoding the result (for non-`id-*` digest algorithm values).
```

hamishwillee commented on files/en-us/web/http/headers/digest/index.md:
```diff
@@ -40,18 +46,20 @@ Digest: <digest-algorithm>=<digest-value>,<digest-algorithm>=<digest-value>
-  - : The result of applying the digest algorithm to the resource representation and encoding the result.
-    The choice of digest algorithm also determines the encoding to use: for example `SHA-256` uses base64 encoding.
+  - : The result of applying the digest algorithm to the resource representation and encoding (for non-`id-*` digest algorithm values) the result.
+    The choice of digest algorithm also determines the encoding to use: for example SHA-256 uses base64 encoding whilst unixsum is represented by a decimal integer.
```
```suggestion
    The choice of digest algorithm also determines the encoding to use: for example SHA-256 uses base64 encoding, whilst unixsum is represented by a decimal integer.
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee commented on files/en-us/web/http/headers/want-repr-digest/index.html:
```diff
@@ -0,0 +1,25 @@
+  - experimental
+---
+
+The **`Want-Repr-Digest`** request or response header states the wish for a {{HTTPHeader("Repr-Digest")}} header.
+
+{{HTTPSidebar}}{{SeeCompatTable}}
```
```suggestion
{{HTTPSidebar}}{{SeeCompatTable}}

The **`Want-Repr-Digest`** request or response header states the wish for a {{HTTPHeader("Repr-Digest")}} header.
```

hamishwillee committed:
Update files/en-us/web/http/headers/want-repr-digest/index.html

hamishwillee commented on files/en-us/web/http/headers/content-digest/index.html:
```diff
@@ -0,0 +1,28 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}
+
+The **`Content-Digest`** response or request header provides a {{Glossary("digest")}} of the _actual message content_, the preimage of computing the _selected representation_ (e.&hairsp;g. [Range Requests](/en-US/docs/Web/HTTP/Range_requests) or [Content Encoding](/en-US/docs/Web/HTTP/Content_negotiation)).  
```
Avoid "e.g.". You'll see it a bit, but we tend to expand out to For example because screen readers do a poor job with these kinds of abbreviation. We also use highlighting sparingly.

```suggestion
The **`Content-Digest`** response or request header provides a {{Glossary("digest")}} of the actual message content, the preimage of computing the _selected representation_.
For example, [Range Requests](/en-US/docs/Web/HTTP/Range_requests) or [Content Encoding](/en-US/docs/Web/HTTP/Content_negotiation).  
```

hamishwillee committed:
Update files/en-us/web/http/headers/content-digest/index.html

hamishwillee commented on files/en-us/web/http/headers/content-digest/index.html:
```diff
@@ -0,0 +1,29 @@
+
+The **`Content-Digest`** response or request header provides a {{Glossary("digest")}} of the actual message content, the preimage of computing the _selected representation_.
+For example, [Range Requests](/en-US/docs/Web/HTTP/Range_requests) or [Content Encoding](/en-US/docs/Web/HTTP/Content_negotiation).  
+Note that the selected representation is independent of wire transmission (e.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}). Note furthermore that the actual message content is independent of [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation): this means that if e.&hairsp;g. a target resource is available in two different MIME types, their correspoding {{HTTPHeader("Repr- ...
```
```suggestion
Note that the selected representation is independent of wire transmission, such as HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}).
Note furthermore that the actual message content is independent of [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation): this means that if a target resource is available in two different MIME types (say), their corresponding {{HTTPHeader("Repr-Digest")}}s without using [Range Requests](/en-US/docs/Web/HTTP/Range_requests) may *both* not match `Content-Digest`, since the "true" content source is never sent.  
```

hamishwillee committed:
Update files/en-us/web/http/headers/content-digest/index.html

hamishwillee commented on files/en-us/web/http/headers/want-content-digest/index.html:
```diff
@@ -0,0 +1,25 @@
+  - experimental
+---
+
+The **`Want-Content-Digest`** request or response header states the wish for a {{HTTPHeader("Content-Digest")}} header.
```
Is this really a response header? (asking, not stating). I should check spec, but I have other things on my queue today.

> 
```suggestion
The **`Want-Content-Digest`** request header indicates a desire for the response to include a {{HTTPHeader("Content-Digest")}} header.
```

hamishwillee commented on files/en-us/web/http/headers/want-content-digest/index.html:
```diff
@@ -0,0 +1,25 @@
+  - experimental
+---
+
+The **`Want-Content-Digest`** request or response header states the wish for a {{HTTPHeader("Content-Digest")}} header.
+
+{{HTTPSidebar}}{{SeeCompatTable}}
```
```suggestion
{{HTTPSidebar}}{{SeeCompatTable}}

The **`Want-Content-Digest`** request or response header states the wish for a {{HTTPHeader("Content-Digest")}} header.
```

hamishwillee commented on files/en-us/web/http/headers/repr-digest/index.html:
```diff
@@ -0,0 +1,49 @@
+  - experimental
+---
+
+The **`Repr-Digest`** response or request header is the `Repr-` analogue to {{HTTPHeader("Content-Digest")}}.
+
+**`Repr-Digest`** takes on an [RFC8941 dictionary](https://www.rfc-editor.org/rfc/rfc8941#section-3.2) whose keys are digest algorithms and whose values for `sha-256` and `sha-512` are [byte sequences](https://www.rfc-editor.org/rfc/rfc8941#section-3.3.5) (colon-wrapped standard base64 with padding).
+
+## Digest algorithms
+
+Only two digest algorithms are currently registered and not considered insecure: `sha-512` and `sha-256`.  
+Insecure digest algorithms registered are: `md5`, `sha` (SHA-1), `unixsum`, `unixcksum`, `adler` (ADLER32) and `crc32c`
+
+{{HTTPSidebar}}{{SeeCompatTable}}
```
```suggestion
{{HTTPSidebar}}{{SeeCompatTable}}

The **`Repr-Digest`** response or request header is the `Repr-` analogue to {{HTTPHeader("Content-Digest")}}.

**`Repr-Digest`** takes on an [RFC8941 dictionary](https://www.rfc-editor.org/rfc/rfc8941#section-3.2) whose keys are digest algorithms and whose values for `sha-256` and `sha-512` are [byte sequences](https://www.rfc-editor.org/rfc/rfc8941#section-3.3.5) (colon-wrapped standard base64 with padding).

## Digest algorithms

Only two digest algorithms are currently registered and not considered insecure: `sha-512` and `sha-256`.  
The registered insecure digest algorithms are: `md5`, `sha` (SHA-1), `unixsum`, `unixcksum`, `adler` (ADLER32) and `crc32c`
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee commented on files/en-us/web/http/headers/want-repr-digest/index.html:
```diff
@@ -0,0 +1,25 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}
+
+The **`Want-Repr-Digest`** request or response header states the wish for a {{HTTPHeader("Repr-Digest")}} header.
```
Is this really a response header (as well)? (asking, not stating). I should check spec, but I have other things on my queue today.

```suggestion
The **`Want-Content-Digest`** request header indicates a desire for the response to include a {{HTTPHeader("Content-Digest")}} header.
```

jfrech replied:
Yes; I am currently also not quite sure what its use would be:
> Senders can indicate their interest in Integrity fields and hashing
> algorithm preferences using the Want-Content-Digest or Want-Repr-
> Digest fields. These can be used in **both requests and responses**.

Cf. [draft-12, 4. Integrity preference fields](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-12#name-integrity-preference-fields) (emphasis mine)

hamishwillee replied:
Thanks for checking. I've emailed the spec authors.

hamishwillee commented on files/en-us/web/http/headers/want-repr-digest/index.html:
```diff
@@ -0,0 +1,25 @@
+{{HTTPSidebar}}{{SeeCompatTable}}
+
+The **`Want-Repr-Digest`** request or response header states the wish for a {{HTTPHeader("Repr-Digest")}} header.
+
```
Examples? Are there options to this - in the same way as `Want-Digest`?

jfrech replied:
Cf. 4aa860481ca

hamishwillee commented on files/en-us/web/http/headers/want-digest/index.md:
```diff
@@ -17,7 +21,7 @@ If `Want-Digest` does not include any digest algorithms that the server supports
-See the page for the {{HTTPHeader("Digest")}} header for more information.
+See also the {{HTTPHeader("Digest")}} header.
 
 <table class="properties">
```
These tables need to appear in your new documents, as do the Syntax and directive boxes. See https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/HTTP_header_page_template

jfrech replied:
Done in a18af4f1aec639ae6e0208ab966f0e31e43dff10; do you agree these header names are not forbidden?

hamishwillee replied:
Yes. Most new ish forbidden header names are prefixed with Sec, and there is nothing I see in the spec to indicate this is forbidden.

hamishwillee commented on files/en-us/web/http/headers/content-digest/index.html:
```diff
@@ -0,0 +1,30 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}
+
+The **`Content-Digest`** response or request header provides a {{Glossary("digest")}} of the actual message content, the preimage of computing the _selected representation_.
```
1. See my comment above about the page template.
2. I do not understand this term. I.e. what is "the preimage ...". I think you are suggesting that that there is some notional "default representation" or "data representation" that we refer to as the "content". So this is the digest of the resource if you don't specify the representation you want? Is that right? We should use common English to explain the differences between this and the representation.

hamishwillee replied:
@jfrech 

So the way I read the spec - now I have done so - is that this is the digest of the actual message content and has nothing to do with the selected representation, unless the selected representation is the actual message content. 

In other words, say you have a resource that is sent across a bunch of messages. This is the digest of just this part of the message. It allows you to verify the integrity of each part of the message independently. 

Does that make sense? So it isn't a "preimage" or whatever

hamishwillee commented on files/en-us/web/http/headers/repr-digest/index.html:
```diff
@@ -0,0 +1,49 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}
+
+The **`Repr-Digest`** response or request header is the `Repr-` analogue to {{HTTPHeader("Content-Digest")}}.
```
So no, based on https://github.com/mdn/content/pull/26885/files#r1209609608 this is not true.

The `Content-Digest` is a digest of _each_ message payload, independent of whether or not the data is part of a multi-part message. It allows integrity checking of a single message, in the same way as a CRC might on a packet.
The `Repr-Digest` is  a Digest of the whole resource, which might be sent across multiple messages. This allows the whole resource to be checked for integrity at the end, after all the parts have been recieved and been reassembled. So if your representation can be sent in a single message, then it might match the `Content-Digest`.

Wihtout digging, I see the `Repr` versions as pretty much the same as the original versions, but with better names that allow them to be differentiated from Content-Digest. 

See comment above about page templates. This one is not bad, but needs syntax block etc.

jfrech replied:
I did have a think; see further down.

hamishwillee commented:
> @hamishwillee May I double-check my interpretation of the spec with you?
> 
> I read it to claim the `Repr-` to be the _Selected Representation_, fully dependent on Content Negotiation which takes `Content-Type` into account. Does this mean `Content-Digest` is of no use whatsoever if the "actual message content" (cf. https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-12#section-2) is not selectable as a representation (e. g. MDN's source files only being selectable as HTML)?

But as per the notes I just added, I think you understand `Content-Digest` to mean "a digest of some default representation of the content". My interpretation is that it means "message-specific digest". That may make your question meaningless. Have a think and then get back to me.

hamishwillee commented:
@jfrech Can you please rename the markdown files from index.html to index.md - these are the ones for which no report is being generated: https://github.com/mdn/content/pull/26885#issuecomment-1556319186

jfrech committed:
glossary: removed "Content-Location"

jfrech committed:
glossary: `s!the target resource sent in an HTTP message body!a resource!`

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

jfrech committed:
glossary: s!the!one!

jfrech committed:
renamed *.html to *.md

jfrech committed:
added properties table ; fattened up stubs

jfrech committed:
fix: interpretation of "content"

jfrech committed:
glossary: fixed "content" interpretation

jfrech commented:
> (...) Have a think and then get back to me.

I had a think and now do agree with you. I thought `Content-` was stronger, but this is not the case.

jfrech committed:
fixed deprecated headers "Digest" and "Want-Digest"

jfrech commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -7,14 +7,13 @@ page-type: glossary-definition
 
 Representations are different forms of a particular resource.  
-For example, the same data might be formatted as a particular media type such as XML or JSON, localized to a particular written language or geographical region, and/or compressed or otherwise encoded for transmission. The underlying resource is the same in each case, but its representation is different.
+For example, the same data might be formatted as a particular media type such as XML or JSON<!--,--> or localized to a particular written language or geographical region<!--, and/or compressed or otherwise encoded for transmission-->. The underlying resource is semantically the same in each case (although details may vary: an e.&hairsp;g. text/html page is inherently more feature-rich than an e.& ...
```
@hamishwillee This is a tad tricky: the deprecated "Digest" header depended on e.&hairsp;g. "Content-Encoding". But "selected representation" in the "Repr-Digest"-sense does not depend on "Content-Encoding"; "Content-Digest", however, does.

jfrech replied:
Ok.

jfrech committed:
pleasing the linter

jfrech committed:
content harmony

jfrech committed:
glossary: added examples

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,61 @@ slug: Glossary/Representation_header
 
-Representation headers may be present in both HTTP request and response messages.
-If sent as a response to a `HEAD` request, they describe the body content that _would_ be selected if the resource was actually requested.
+Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms: an HTTP message frames (cf. e.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}) a particular stream of octets (cf. e.&hairsp;g. {{HTTPHeader("Content-Range")}}) derived from the _selected representation_.
```
Can you globally purge the use of `&hairsp;`?

We might add a comment here about the fact that in some case `<!--, and/or compressed or otherwise encoded for transmission-->` is relevant?

hamishwillee replied:
P.S. It doesn't quite "fit", but could perhaps be a "Note" since it is related, but also an aside.

hamishwillee commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -4,23 +4,61 @@ slug: Glossary/Representation_header
+
+Representation headers include: {{HTTPHeader("Content-Type")}}, {{HTTPHeader("Content-Encoding")}}, {{HTTPHeader("Content-Language")}}, {{HTTPHeader("Content-Range")}}, {{HTTPHeader("Repr-Digest")}}, and {{HTTPHeader("Digest")}} {{Deprecated_Inline}}.
+
+## Examples
```
Actually, I was wrong sorry. We try and avoid having examples in glossary docs. We try to keep the glossary as short as possible. These examples should exist, if they do not, in the reference pages - e.g. want-repr-digest.

jfrech replied:
It didn't fit in the glossary. See bc31ed5466175c2e3f10fb759779cb43ddfecca2.

jfrech committed:
no examples in glossaries

hamishwillee commented on files/en-us/web/http/headers/content-digest/index.md:
```diff
@@ -0,0 +1,65 @@
+
+## Directives
+
+For permissible digest algorithms see {{HTTPHeader("Repr-Digest")}}.
```
I would reproduce them. This is arguable of course. I just like making a tiny bit easier for readers. You could still say "this is the same set as in Repr-Digest"

hamishwillee commented on files/en-us/web/http/headers/want-content-digest/index.md:
```diff
@@ -0,0 +1,55 @@
+
+## Directives
+
+For permissible digest algorithms see {{HTTPHeader("Repr-Digest")}}.
```
Duplicate?

jfrech replied:
I don't think so; there are four files in total:
```
patch-1 $ grep -riI 'For permissible digest'
files/en-us/web/http/headers/want-repr-digest/index.md:For permissible digest algorithms see {{HTTPHeader("Repr-Digest")}}.
files/en-us/web/http/headers/content-digest/index.md:For permissible digest algorithms see {{HTTPHeader("Repr-Digest")}}.
files/en-us/web/http/headers/want-content-digest/index.md:For permissible digest algorithms see {{HTTPHeader("Repr-Digest")}}.
```

hamishwillee commented on files/en-us/web/http/headers/want-repr-digest/index.md:
````diff
@@ -0,0 +1,55 @@
+
+## Examples
+
+```http
````
So the request and response examples you put in the glossary would go better here - expanding out what a user might see. Ditto for the others.

jfrech replied:
I put them in "Repr-Digest" (bc31ed5466175c2e3f10fb759779cb43ddfecca2), since I feel like the three others are variants.

hamishwillee replied:
We tend to duplicate the reciprocal pairs; so you might have the same example in both. But hold off doing anything until Teoli comments. I might not be up to date on current patterns in this area.

jfrech committed:
Update files/en-us/glossary/representation_header/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

jfrech committed:
minor: capitalised "Markdown"

hamishwillee commented:
@jfrech I think we're getting there. I'll see if I can get @teoli2003 to do subsequent review as he is more expert than me. He's been a little silent this last week or two so might be on holiday.

jfrech committed:
[wip] "Content-Digest"

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

jfrech committed:
"Content-Digest"

jfrech commented on files/en-us/web/http/headers/content-digest/index.md:
```diff
@@ -11,8 +11,8 @@ status:
 
 The **`Content-Digest`** response or request header provides a {{Glossary("digest")}} of the actual message content, the stream of octets framed in an HTTP message.
-As such, `Content-Digest` is dependent on the {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}} , but not dependent on, for example, HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}.
-`Content-Digest` would have the same value as {{HTTPHeader("Repr-Digest")}} if a representation was sent in a single message.
```
@hamishwillee My problem with "would have the same value as" is that first of all *sameness* could only apply in the dictionary sense and furthermore, e.&hairsp;g.
```http
Content-Digest: sha512=:z4PhNX7vuL3xVChQ1m2AB9Yg5AULVxXcg/SpIdNs6c5H0NE8XYXysP+DGNKHfuwvY7kxvUdBeoGlODJ6+SfaPg==:
Content-Digest: sha256=:47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=:
```
both describe the empty file yet there is (by design) no known notion of *sameness*.

hamishwillee replied:
Fair enough! Reword as you think appropriate.

jfrech committed:
anticipating merge conflict

jfrech committed:
anticipating merge conflict

jfrech commented:
I will want to squash first before merging with master. @hamishwillee should I do?

jfrech committed:
note on Content Negotiation

jfrech committed:
note on not using insecure digest algorithms

hamishwillee commented:
We'll squash as part of the merge :-)

I guess @teoli2003 is still on holiday. If we get no response by Wednesday next week I'll do another review and merge.

hamishwillee commented on files/en-us/web/http/headers/repr-digest/index.md:
```diff
@@ -10,7 +10,7 @@ status:
 {{HTTPSidebar}}{{SeeCompatTable}}
 
-The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.&hairsp;g. {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}.
+The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.&hairsp;g. {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}. Note furthermore that [Content Negotiation](/en- ...
```
Remove all cases of `e.hairsp;`

jfrech replied:
Did you mean to replace my subphrase?

hamishwillee replied:
No. 

What I am saying here is that I don't see the benefit of ``e.hairsp;g.` over `e.g.`

hamishwillee commented on files/en-us/web/http/headers/repr-digest/index.md:
```diff
@@ -0,0 +1,134 @@
+
+## Syntax
+
+`Repr-Digest` describes an [RFC8941 dictionary](https://www.rfc-editor.org/rfc/rfc8941#section-3.2) with its keys being names of digest algorithms and its values being the digest in bytes (or an integer digest for legacy digest algorithms).
+
+> **Note:** In contrast to earlier drafts of the specification, the standard-base64-encoded digest bytes are wrapped in colons (`:`, ASCII 0x3A) as part of the [dictionary syntax](https://www.rfc-editor.org/rfc/rfc8941#name-byte-sequences).
```
Does not match the template https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/HTTP_header_page_template

Unless absolutely necessary, do not link the spec. It's our job to show how the API is used "for humans".

```suggestion

`Repr-Digest` directives dictionaries, with keys being names of digest algorithms and values being the digest in bytes, wrapped in colons (`:`, ASCII 0x3A).

```http-nolint
# Secure digest algorithms
Repr-Digest sha512=:calculated_digest:
Repr-Digest: sha-256=:calculated_digest:

# Insecure digest algorithms
Repr-Digest md5=:calculated_digest:
Repr-Digest sha=:calculated_digest:
Repr-Digest unixsum=:calculated_digest:
Repr-Digest unixcksum=:calculated_digest:
Repr-Digest adler=:calculated_digest:
Repr-Digest crc32c=:calculated_digest:
```

> 
(or an integer digest for legacy digest algorithms).
The 
> **Note:** Legacy digest algorithms (in earlier drafts of the specification), use an integer digest and do nor wrap it in colons.
```

jfrech replied:
I think it is much more human-readable if one does not say "a dictionary" (which is very ambiguous) but rather "an RFC8941 dictionary" which specifically states which syntactical form of a dictionary is at play here.

Furthermore, half-correctly and vaguely describing the dictionary as you have done benefits no one.

teoli2003 replied:
I wonder if we should have a glossary entry for RFC8941 dictionary: I don't like to send readers to the spec (that are difficult to understand) to grasp the syntax.

hamishwillee commented on files/en-us/web/http/headers/repr-digest/index.md:
````diff
@@ -0,0 +1,140 @@
+  </tbody>
+</table>
+
+## Syntax
+
+```http
+Repr-Digest <directive_dictionary>
+```
+
+## Directives
+
+`Repr-Digest` directives (`<directive_dictionary>`) are dictionaries with keys being names of digest algorithms and values being the digest in bytes, wrapped in colons (`:`, ASCII 0x3A).
+For example:
+
+```http
+Repr-Digest: sha-256=:AEGPTgUMw5e96wxZuDtpfm23RBU3nFwtgY5fw4NYORo=:
+```
+
+The registered algorithms that are considered secure are: `sha-512` and `sha-256`.
+
+The registered algorithms that are considered insecure are: `md5`, `sha` (SHA-1), `unixsum`, `unixcksum`, `adler` (ADLER32) and `crc32c`.
+
+> **Note:** Using insecure digest algorithms is discouraged, because collisions can realistically be forced (making the digests less useful).
+> It is often better to not to emit a `Repr-Digest` at all than to emit one that specifies an insecure digest algorithm.
````
FYI, I updated this to try put in format acceptable to the template. Let's see what @teoli2003 says.

jfrech replied:
Your statement is incorrect: e. g. "unixcksum" is not a byte array dictionary value.

hamishwillee commented:
@jfrech Other than the linter nits I think I've taken this as far as I want to. @teoli2003 should be around soon, and hopefully will take this to the end.

teoli2003 commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -6,23 +6,28 @@ page-type: glossary-definition
 
-Representation headers may be present in both HTTP request and response messages.
-If sent as a response to a `HEAD` request, they describe the body content that _would_ be selected if the resource was actually requested.
+Whilst representations are different forms of resources, representations can themselves also be transmitted in different forms: an HTTP message frames (cf. e.&hairsp;g. HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}) a particular stream of octets (cf. e.&hairsp;g. {{HTTPHeader("Content-Range")}}) derived from the _selected representation_.
```
```suggestion
While representations are different forms of resources, representations can themselves also be transmitted in various forms: an HTTP message frames (cf., e.g., HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}), a particular stream of octets (cf., e.g., {{HTTPHeader("Content-Range")}}) derived from the _selected representation_.
```

teoli2003 commented on files/en-us/glossary/representation_header/index.md:
```diff
@@ -6,23 +6,28 @@ page-type: glossary-definition
 - {{glossary("Entity header")}}
-- {{HTTPHeader("Digest")}}/ {{HTTPHeader("Want-Digest")}}
+- {{HTTPHeader("Repr-Digest")}}, {{HTTPHeader("Want-Repr-Digest")}}
+- pertaining to particular forms of the selected representation: {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}
```
```suggestion
- {{HTTPHeader("Content-Digest")}}, {{HTTPHeader("Want-Content-Digest")}}
```

teoli2003 commented on files/en-us/web/http/headers/content-digest/index.md:
```diff
@@ -0,0 +1,67 @@
+As such, `Content-Digest` is dependent on among other things {{HTTPHeader("Content-Encoding")}} and {{HTTPHeader("Content-Range")}}, but not dependent on, for example, HTTP/1.1's {{HTTPHeader("Transfer-Encoding")}}.
+`Content-Digest` may coincide with {{HTTPHeader("Repr-Digest")}} if a representation was sent in a single message.
+
+"content" in this setting refers to a particular octet representation of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource.
```
```suggestion
In this setting, _content_ refers to a particular octet representation of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource.
```

teoli2003 commented on files/en-us/web/http/headers/repr-digest/index.md:
```diff
@@ -0,0 +1,134 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}
+
+The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.&hairsp;g. {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}. Note furthermore that [Content Negotiation](/en- ...
```
```suggestion
The **`Repr-Digest`** response or request header provides a {{Glossary("digest"}} of the [selected representation](https://www.rfc-editor.org/rfc/rfc9110#section-6.4) of the target resource. It is invariant under e.g., {{HTTPHeader("Content-Encoding")}} or {{HTTPHeader("Content-Range")}}, which do affect the {{HTTPHeader("Content-Digest")}}. Furthermore, [Content Negotiation](/en-US/docs/Web/HTTP/Content_negotiation) can result in different selected representations with different representation digests.
```

teoli2003 commented on files/en-us/web/http/headers/repr-digest/index.md:
```diff
@@ -0,0 +1,134 @@
+
+## Examples
+
+### Example: HTTP response where `Repr-Digest` and `Content-Digest` coincide
```
```suggestion
### HTTP response where `Repr-Digest` and `Content-Digest` coincide
```

teoli2003 commented on files/en-us/web/http/headers/repr-digest/index.md:
````diff
@@ -0,0 +1,134 @@
+...
+```
+
+### Example: HTTP responses where {{HTTPHeader("Repr-Digest")}} and {{HTTPHeader("Content-Digest")}} diverge
````
```suggestion
### HTTP responses where {{HTTPHeader("Repr-Digest")}} and {{HTTPHeader("Content-Digest")}} diverge
```

teoli2003 commented on files/en-us/web/http/headers/repr-digest/index.md:
````diff
@@ -0,0 +1,134 @@
+...
+```
+
+### Example: Successful HTTP request-response employing {{HTTHeader("Want-Repr-Digest")}}, {{HTTHeader("Repr-Digest")}}, {{HTTHeader("Content-Digest")}}
````
```suggestion
### Successful HTTP request-response employing {{HTTHeader("Want-Repr-Digest")}}, {{HTTHeader("Repr-Digest")}}, {{HTTHeader("Content-Digest")}}
```

teoli2003 commented on files/en-us/web/http/headers/repr-digest/index.md:
````diff
@@ -0,0 +1,134 @@
+...
+```
+
+### Example: Unsuccessful HTTP request-response employing {{HTTHeader("Repr-Digest")}}
````
```suggestion
### Unsuccessful HTTP request-response employing {{HTTHeader("Repr-Digest")}}
```

teoli2003 commented on files/en-us/web/http/headers/want-content-digest/index.md:
```diff
@@ -0,0 +1,55 @@
+
+`Want-Content-Digest` describes an [RFC8941 dictionary](https://www.rfc-editor.org/rfc/rfc8941#section-3.2) with its keys being hashing algorithms and its values being the integers `0` (meaning "not acceptable") or `1` to `9` (conveying ascending, relative, weighted preference).
+
+> **Note:** In contrast to earlier drafts of the specifications, the weighting is *not* declared via [q-values](/en-US/docs/Glossary/Quality_values).
```
```suggestion
> **Note:** In contrast to earlier drafts of the specifications, the weighting is _not_ declared via [q-values](/en-US/docs/Glossary/Quality_values).
```

teoli2003 commented on files/en-us/web/http/headers/want-repr-digest/index.md:
```diff
@@ -0,0 +1,55 @@
+
+`Want-Repr-Digest` describes an [RFC8941 dictionary](https://www.rfc-editor.org/rfc/rfc8941#section-3.2) with its keys being hashing algorithms and its values being the integers `0` (meaning "not acceptable") or `1` to `9` (conveying ascending, relative, weighted preference).
+
+> **Note:** In contrast to earlier drafts of the specifications, the weighting is *not* declared via [q-values](/en-US/docs/Glossary/Quality_values).
```
```suggestion
> **Note:** In contrast to earlier drafts of the specifications, the weighting is _not_ declared via [q-values](/en-US/docs/Glossary/Quality_values).
```

teoli2003 reviewed (changes requested):
A few extra points, but overall this looks good and close to be ready

bsmth commented:
Hi @jfrech - thanks a lot for the submission. There are some review comments outstanding above, would you like to come back to this one? @teoli2003 has indicated that ["this looks good and close to be ready"](https://github.com/mdn/content/pull/26885#pullrequestreview-1532019996), so it would be great to have your work land eventually.

wbamberg commented:
@bsmth , except for https://github.com/mdn/content/pull/26885#discussion_r1264900105, @teoli2003 has provided suggestions for all his comments. Perhaps we could just accept them all and merge this, and let https://github.com/mdn/content/pull/26885#discussion_r1264900105 go?

bsmth commented:
> @bsmth , except for [#26885 (comment)](https://github.com/mdn/content/pull/26885#discussion_r1264900105), @teoli2003 has provided suggestions for all his comments. Perhaps we could just accept them all and merge this and let [#26885 (comment)](https://github.com/mdn/content/pull/26885#discussion_r1264900105) go?

That sounds fine (I see we added the awaiting resp label `on Dec 8, 2023`) let's try to get the improvements merged, in that case. I don't think the author is active on GitHub anymore by the looks of it.

bsmth committed:
Apply suggestions from code review

Co-authored-by: Jean-Yves Perrier <jypenator@gmail.com>

bsmth committed:
Merge branch 'main' into patch-1

wbamberg committed:
Fix macro invocation errors

wbamberg committed:
Fix front matter, maybe

wbamberg commented:
OK, @bsmth , I have fixed the macro invocation errors and the front matter linting errors. There are still reviewdog comments but _I think_ we can ignore them.

I haven't tried to properly review the PR content but from the existing reviews it seems close and surely better than what we have at the moment. Maybe we can give @hamishwillee and @teoli2003 a couple of days to re-review, and if there's no response merge it then?

jfrech commented:
@wbamberg It looks like the linter only complained about ordering, which you seemed to have abided by. Resolving its conversations.

jfrech commented on files/en-us/web/http/headers/repr-digest/index.md:
```diff
@@ -43,7 +43,7 @@ Usage of digest algorithms which are considered insecure is discouraged as colli
 
 ## Examples
 
-### HTTP response where `Repr-Digest` and `Content-Digest` coincide
```
Is markup not allowed in headers or is this edit of your own taste? Looking at <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers>, HTTP headers seem to be marked up as inline code.

wbamberg replied:
Markup is allowed. tbh all that happened here was that I removed the macro calls from some of the headers, which also removed the code style, then I removed it here just to be consistent. I don't mind either way, and if you'd like to put the code style back (not the macros though) I am happy with that.

jfrech replied:
Done in bf6ba95d087150a1b413dd1363b28069b9bc8bc8.

jfrech committed:
http headers (Want-)*(Repr|Content)-Digest: markup of verbatim HTTP headers

hamishwillee reviewed (approved):
I'm away so I can't give it my full attention, but it looks good to me on scan (and it's had a lot of eyes on it).

teoli2003 reviewed (approved):
👍
